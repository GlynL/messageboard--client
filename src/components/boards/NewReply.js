import React, { Component } from "react";
import { connect } from "react-redux";
import { compose } from "redux";
import { withRouter } from "react-router-dom";
import * as threadActions from "../../actions/threads";
import * as boardActions from "../../actions/boards";

const INITIAL_STATE = {
  text: ""
};

class NewReply extends Component {
  constructor(props) {
    super(props);
    this.state = INITIAL_STATE;
  }

  handleSubmit = e => {
    e.preventDefault();
    const { board, thread } = this.props.match.params;
    this.props.addReply(
      this.state,
      board,
      thread,
      this.props.token,
      this.props.updateBoard
    );
    this.setState(INITIAL_STATE);
  };

  handleChange = e => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  render() {
    if (!this.props.token) return <div>Login to leave a reply.</div>;
    return (
      <div>
        <h2>Add your reply</h2>
        <p>{this.props.error}</p>
        <form onSubmit={this.handleSubmit}>
          <label htmlFor="text">Text</label>
          <textarea
            onChange={this.handleChange}
            value={this.state.text}
            name="text"
            id="text"
          />
          <button>Add Reply</button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  token: state.auth.authenticated,
  error: state.threads.errorMessage
});

export default compose(
  withRouter,
  connect(
    mapStateToProps,
    { ...threadActions, ...boardActions }
  )
)(NewReply);
