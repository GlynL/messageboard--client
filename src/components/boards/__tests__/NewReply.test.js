import React from "react";
import { shallow } from "enzyme";
import { NewReply } from "../NewReply";

it("matches snapshot w/out auth", () => {
  const wrapper = shallow(<NewReply />);
  expect(wrapper).toMatchSnapshot();
});

describe("w/ auth", () => {
  let wrapper, addReply, match;
  beforeEach(() => {
    match = {
      params: {
        board: "test board",
        thread: "test thread"
      }
    };
    addReply = jest.fn();
    wrapper = shallow(
      <NewReply token="test token" addReply={addReply} match={match} />
    );
  });

  it("matches snapshot w auth", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("textarea & form", () => {
    beforeEach(() => {
      wrapper
        .find("textarea")
        .simulate("change", { target: { value: "test", name: "text" } });
      wrapper.update();
    });

    it("handles textarea change", () => {
      expect(wrapper.find("textarea").prop("value")).toBe("test");
      expect(wrapper.state("text")).toBe("test");
    });

    it("handles form submit", () => {
      const preventDefault = jest.fn();
      wrapper.find("form").simulate("submit", { preventDefault });
      wrapper.update();
      expect(preventDefault).toHaveBeenCalled();
      expect(addReply).toHaveBeenCalled();
      expect(wrapper.find("textarea").prop("value")).toBe("");
      expect(wrapper.state("text")).toBe("");
    });
  });
});

it("matches snapshot w/ auth & error", () => {
  const wrapper = shallow(<NewReply token="test token" error="test error" />);
  expect(wrapper).toMatchSnapshot();
});
