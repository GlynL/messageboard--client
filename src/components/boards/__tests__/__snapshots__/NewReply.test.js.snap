// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot w/ auth & error 1`] = `
<div>
  <h2>
    Add your reply
  </h2>
  <p>
    test error
  </p>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="text"
    >
      Text
    </label>
    <textarea
      id="text"
      name="text"
      onChange={[Function]}
      value=""
    />
    <button>
      Add Reply
    </button>
  </form>
</div>
`;

exports[`matches snapshot w/out auth 1`] = `
<div>
  Login to leave a reply.
</div>
`;

exports[`w/ auth matches snapshot w auth 1`] = `
<div>
  <h2>
    Add your reply
  </h2>
  <p />
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="text"
    >
      Text
    </label>
    <textarea
      id="text"
      name="text"
      onChange={[Function]}
      value=""
    />
    <button>
      Add Reply
    </button>
  </form>
</div>
`;
